### ISSUE TITLE
Spike: Add LB real time capabilities to Angular models

### ISSUE THREAD

**Issue by altsang** [strongloop/loopback-sdk-angular-cli#23]
As a LB front end developer using Angular, I would want to be able to use the Angular SDK CLI tool to offer the scaffolding capability to create Angular models that are sync'ed back to LoopBack using loopback-pubsub with loopback-replication so that I have models and views updated in realtime for both front end and backend.

**Comment by chandadharap** [comment]
Outcome: Understand requirements, tech design eval, break up estimated stories.

**Comment by chandadharap** [comment]
stretch goal: demoable

**Comment by ritch** [comment]
# LoopBack-Angular Sync

### Requirements

The ability to sync collections of objects between an angular application
and a loopback datasource as data changes (on the client on in the datasource).

The client should not have to constantly check the server for changes, instead
the server should tell the client when new data is available.

Useable out of the box by any angular app in a similar way that the loopback-sdk is.

**Comment by ritch** [comment]
### Changes

**New LoopBack Standalone Client**

 - Supports angular digest / ngHttp (instead of request)
 - Static file (users do not need to build their app, just export an api definition)
 - Can be used without angular

```txt
slc loopback:export

{
  MyModel: {
    prototype: {
      find: {
        http: {path: '/find'}
        accepts: [],
        returns: []
      }
    }
  }
}

wrote file export.json

slc loopback:browser-client export.json --angular

including module xyz 85kb (45kb min.)
including module abc 85kb (5kb min.)
including module def 85kb (45kb min.)
including module jkl 85kb (3kb min.)

wrote client.js (size: 108kb (88kb min))
```

```js
// in browser
var app = new LoopBackClient('http://localhost:3000');

app.models.MyModel.find(console.log); // => [...MyModel, ...]
```

```txt
slc loopback:angular export.json
```

```js
// in angular app
// include the static file "loopback-angular-service.js"
function myCtlr($scope, LoopBackClient) {
  var app = new LoopBackClient('/api', {
    definition: def // from exports.json
  });

  // when loopback runs within an angular app
  // it behaves like ngResources by calling
  // $rootScope.$apply() after a method is invoked
  $scope.myModels = app.models.MyModel.find();
}
```

**changes required to do this**
- support angular as a context by checking `window && window.angular`
- use angular.getCurrentScope() or similar when in angular context
to call scope.apply()
- or patch promises in a angular compatible way
- support promises in remote methods
- support invoking methods and expecting a promise
- support using ng-http for making requests instead of request when available
- support exporting serialized version of remotes (mostly there)
- support building remotes object from serialized version
- implement slc tooling
  - slc loopback:client 
   - uses browserify and the exported remotes.json file to build a client bundle
   - that has 0 deps on browserify, angular
   - only includes client code (no server / isomorphic code included)
   - client includes validation / replication an other features from loobpack
   - single api for browser apps (including angular apps) and node
    - no more confusing MyModel.find({filter: ...}) and MyModel.$prototype$updateAttr()
    - bug fixes in loopback will fix browser client and server
   - sep. exports.json and code (heavily cache the client.js file and change the exports.json file)

**Comment by ritch** [comment]
@bajtos can you take a look at the idea for a standalone client and comment?

**Comment by ritch** [comment]
If we are ok with the standalone client approach then here is what I think we should do for the rest of the implementation:

```js
var changes = Todo.changes({
  where: {owner: ...}
});

changes.on('data', function(change) {
  origin: 'client id',
  change.previous // the entire previous version
  change.current // the current version

  var todo = getTodoById(change.previous.getId());
  if(todo) {
    // a manual "update"
    todo.text = change.current.text;
    todo.done = change.current.done;
    // this also supports updating ids since we include the previous id
    todo.id = change.current.id;
  } else {
    // it has been deleted from the collection
  }
});


var todos = Todos.createSyncedCollection({
  where: {owner: me}
});

function createSyncedCollection(filter) {
  var collection = new SyncCollection();
  var changes = this.changes(filter);
  changes.on('data', function(change) {
    var target = collection.find(change.getId());

    if(Todos.applyChange) {
      Todos.applyChange(change, target);
    } else {
      SyncCollection.applyChange(change, target);
    }
  });
}

function changes(filter, accessToken) {
  var ps = this.app.getPubSubClient();
  var topic = this.getTopicUrl() + '?filter=' enc(filter) + 'accessToken=' + accessToken ? accessToken : '';

  // the server impl. will need to bridge this topic 
  // with the generic topic `/todos` and publish back down to this
  // topic only the changes that match the filter
  ps.subscribe(topic);
  var changeStream = new ReadStream();
  var buffer = [];

  ps.on('message', function(msg) {
    if(msg.topic === topic) {
      buffer.push(JSON.parse(msg));
    }
  });

  changeStream._read = function(n) {
    var result = [];
    while(n--) {
      result.push(buffer.shift());
    }
    return result;
  }
}


// example usage in an angular controller
function myController($scope, LoopBackClient) {
  var client = LoopBackClient('/api', def);
  var filter = {where: {owner: myUserId}};
  $scope.todos = client.models.Todos.createSyncedCollection(filter);

  $scope.saveTodo = function(todo) {
    todo.save();
  }
}
```

**Comment by ritch** [comment]
### Spike Findings:

There are two types of "sync" that we need to support.

**Immediate Sync**
 - suitable for clients with persistent connections to a server
 - near realtime latency
 - meteor like features (change on one client, see the change immediately on the other)
 - resolve conflicts by applying updates multiple times
 - only track the version a change is being made against (no revisions required)

**Intermediate Sync**
 - suitable for clients who do not have a connection to the server at all times
 - servers and clients must keep journals of latest changes
 - syncs should not be often, only once reconnecting
 - conflicts are likely, users should have a lot of control for resolving conflicts

**Comment by bajtos** [comment]
@ritch Lot's of interesting stuff :+1: I'll need some time to digest the proposed implementation and think through consequences & edge cases.

As for your question whether we should reduce the implementation of the angular sdk by building it on top of the isomorphic client, I feel the current version of the isomorphic client is quite difficult to use, especially compared to the "old" code-generated Angular client. The answer thus really depends on what is the scope of this work (what features we want to provide) and how much time we have to implement them.

See https://github.com/strongloop/loopback-sdk-angular/issues/83 and https://docs.google.com/document/d/1mjBbSdr6ho-wEyB1AyLUhYSPUBQq8WQwwnhBrG00kFg/edit for a list of tasks I considered important to implement to in order make the isomorphic client with offline sync super easy to use.

The scenario "Intermediate Sync" described in your last comment above requires change tracking & replication, which is something that does not make sense to reimplement in the code-generated Angular client IMO. In other words, if our goal is to support "Intermediate Sync", then we need to implement pub-sub in the isomorphic client and the code-generated client won't be supported at all.

On the other hand, I can imagine that "Immediate Sync" can be implemented as a small addition to the current code-generated client. That will allow us to deliver a pub-sub feature much sooner, even though it won't be as powerful as what can be achieved in the isomorphic client.

**Comment by ritch** [comment]
I think we need to support both intermediate sync (for offline use cases) and immediate sync (for (near) realtime use cases).

Also after some more discussions about angular, I think it is possible to implement these and CRUD capabilities without heavily depending on angular. Instead handling angular compatibility when it is present.

**Comment by ritch** [comment]
Spoke to @bajtos and the next step on this spike is to modify the offline-sync example to do immediate sync (it might do this already) and measure the latency (how many requests, what the UX is like, etc).

**Comment by ritch** [comment]
What we need to determine with this spike is whether or no we need an enitrely new sync approach (one outline above).

**Comment by ritch** [comment]
@bajtos so it looks like we cannot build the realtime (immediate sync) on top of replication. 

Lets split apart the requirements for immediate sync and highlight the potential edge cases:

**Getting the initial set of filtered data**
 - what if a change occurs while the data is being fetched?

**Notifying Clients of Changes**
 - latency should be much shorter than polling
 - what if a notification is missed?

Here is some psuedo code implementing the basic idea for notifying clients of changes:

```js
// when a client wants a change stream
// client => server "i am interested in these changes"

// server sets up a client to the broker listening on all changes
// when a change matches the filter, it is published as a msg
// to the client

// browser
client.subscribe('/api/todos?filter[where][owner]=27&access_token=...')

// server
client.before('subscribe', function(topic) {
  var filterTopic = topic;
  var parsedTopic = parseTopic(topic);
  var token = parsedTopic.accessToken;
  var model = registry.getModel(parsedTopic.modelName);
  authenticate(accessToken, model, function(err) {
    var modelChangeTopic = '/changes/' + parsedTopic.modelName;
    bridge.subscribe('/changes/' + parsedTopic.modelName);

    bridge.on('message', function(topic, msg) {
      if(topic == modelChangeTopic) {
        client.publish(filterTopic, msg);
      }
    });
  });
});
```

**Applying Changes**
 - what if a change conflicts with a local change (two users interacting with the same object at the same time)


Anyways we don't have to discuss all here, just setting up our meeting for tomorrow morning.

**Comment by chandadharap** [comment]
Is there a design write up under Spikes on the wiki? Please link here.

**Comment by ritch** [comment]
There isn't a write up. @bajtos and I have already agreed on next steps and they are detailed in several user stories:

https://github.com/strongloop/loopback/issues/1420
https://github.com/strongloop/loopback/issues/1419
https://github.com/strongloop/loopback/issues/1418
